pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java

```diff
@@ -366,6 +360,9 @@ private void doIndividualBatchAckAsync(BatchMessageIdImpl batchMessageId) {
                     return value;
                 });
         bitSet.clear(batchMessageId.getBatchIndex());
+        if (pendingIndividualBatchIndexAcks.size() >= MAX_ACK_GROUP_SIZE) {
+            flush();
+        }
```

Comment:
`doIndividualBatchAck` also invokes `doIndividualBatchAckAsync`, what's the difference?

Reply:
Maybe we could add the logic here because only this method could add entries to `pendingIndividualBatchIndexAcks`.

Reply:
> `doIndividualBatchAck` also invokes `doIndividualBatchAckAsync`, what's the difference?

Maybe the name of the method is a little inappropriate.

Reply:
> Maybe we could add the logic here because only this method could add entries to `pendingIndividualBatchIndexAcks`.

Yes, I know. I mean: we check `pendingIndividualAcks` in the `doIndividualAck`, so we should keep the same code logic.

Reply:
> we check pendingIndividualAcks in the doIndividualAck, so we should keep the same code logic.

It makes sense to me.

Reply:
It seems that there is also a method `PersistentAcknowledgmentsGroupingTracker#addListAcknowledgment` that may increase the pending individual count.